src/idl/tests/annotation.c

```diff
@@ -45,6 +45,196 @@ parse_string(uint32_t flags, const char *str, idl_pstate_t **pstatep)
   return ret;
 }
 
+static void test_optional(const char *parsestring, idl_retcode_t parseresult, bool funcresult)
+{
+  idl_pstate_t *pstate = NULL;
+  idl_retcode_t ret = parse_string(IDL_FLAG_ANNOTATIONS, parsestring, &pstate);
+  CU_ASSERT_EQUAL_FATAL(ret, parseresult);
+
+  if (pstate) {
+    idl_node_t *node;
+    int nstructs = 0;
+    IDL_FOREACH(node, pstate->root) {
+      if (!idl_is_struct(node))
+        continue;
+      nstructs++;
+      idl_struct_t *s = (idl_struct_t *)node;
+      CU_ASSERT_PTR_NOT_NULL_FATAL(s);
+      CU_ASSERT_FATAL(idl_is_struct(s));
+      assert(s);
+      idl_member_t *m = NULL;
+      idl_declarator_t *d = NULL;
+      IDL_FOREACH(m, s->members) {
+        CU_ASSERT_EQUAL(m->optional.value, funcresult);
+        CU_ASSERT_EQUAL(idl_is_optional(m), funcresult);
+        IDL_FOREACH(d, m->declarators) {
+          CU_ASSERT_EQUAL(idl_is_optional(d), funcresult);
+        }
+      }
+    }
+    CU_ASSERT_EQUAL(nstructs, 1);
+    idl_delete_pstate(pstate);
+  }
+}
+
+CU_Test(idl_annotation, optional)
+{
+  static const struct {
+    idl_retcode_t r;
+    bool v;
+    const char *s;
+  } tests[] = {
+    {IDL_RETCODE_OK,                  true,   "struct s { @optional char c; };"                       },
+    {IDL_RETCODE_OK,                  false,  "struct s { @optional(false) char c; };"                },
+    {IDL_RETCODE_OK,                  true,   "struct s { @optional(true) char c; };"                 },
+    {IDL_RETCODE_OK,                  true,   "struct s { @optional(true) char c_1, c_2, c_3; };"     },
+    {IDL_RETCODE_OK,                  true,   "struct s { @optional(true) char c_1, c_2, c_3; };"     },
+    {IDL_RETCODE_OK,                  true,   "struct s { @optional sequence<double> s_d; };"         },
+    {IDL_RETCODE_OK,                  true,   "typedef sequence<long> seq_long;\n"
+                                              "struct s { @optional seq_long s_l_a[15]; };"           },
+    {IDL_RETCODE_SEMANTIC_ERROR,      false,  "@optional struct s { char c; };"                       },
+    {IDL_RETCODE_ILLEGAL_EXPRESSION,  false,  "struct s { @optional(\"true\") char c; };"             },
+    {IDL_RETCODE_ILLEGAL_EXPRESSION,  false,  "struct s { @optional(123) char c; };"                  },
+    {IDL_RETCODE_ILLEGAL_EXPRESSION,  false,  "struct s { @optional(0.1) char c; };"                  },
+    {IDL_RETCODE_SEMANTIC_ERROR,      false,  "enum e {  e_0, @optional e_1};"                        },
+    {IDL_RETCODE_SYNTAX_ERROR,        false,  "union u switch (long) { case 0: @optional double d; };"}
```

Comment:
Same for this one because we're actually testing the syntax rather than something `@optional` specific.

Reply:
Same reasoning as previous comment.